### PR TITLE
Consumption Improvements

### PR DESCRIPTION
**Pull request by sergiosalvatore**
- I was getting "broker not connected" errors frequently when starting up that end up being a race between requesting partition information and having the broker's connection established.  This implements a simply retry+backoff so that it can hopefully succeed.
- I was also having issues with the output being interleaved on account of the multiple goroutines writing to same writer without synchronization (in some places).  I fixed that with a single channel which handles all the writing.  This is mentioned in #23 
- I wonder if we can use sarama's newer [ConsumerGroup](https://godoc.org/github.com/Shopify/sarama#ConsumerGroup) interface which might simplify this -- I have to dig a bit more into that one...

Thoughts welcome... :-)

### CODE REVIEW

File: cmd/kaf/consume.go [hunk 1]
```diff
@@ -91,15 +93,29 @@ var consumeCmd = &cobra.Command{
...
 				if err != nil {
 					panic(err)
 				}
+
 				followOffset := offsets.GetBlock(topic, partition).Offset - 1
 
 				if follow && followOffset > 0 {
 					offset = followOffset
-					fmt.Fprintf(os.Stderr, "Starting on partition %v with offset %v\n", partition, offset)
+					outputChan <- outputMessage{error: fmt.Errorf("starting on partition %v with offset %v", partition, offset)}
```
Comment (mafredri) [inline cmd/kaf/consume.go]: With this change, errors and the output message are no longer synchronized. An alternative approach here would be to append to a slice, say `om.errors = append(om.errors, fmt.Errorf("..."))` and only sending on the output channel once.

**Comment by birdayz** [comment]
1) I think it's a bad idea to retry like this. We should rather try to make sarama fetch broker metadata before we do our own logic

so instead of wrapping GetOffsets, we should do this wayy before, not tied to GetOffset itself. i'm not sure how we can do that, but i should be possible.

2) about interleaved output: there was a recent PR on this, using a mutex - can you elaborate why this is still problematic?

File: cmd/kaf/consume.go (hunk 1 above)
Comment (sergiosalvatore) [reply cmd/kaf/consume.go]: That's a good point...  Should be an easy fix.

**Comment by sergiosalvatore** [comment]
Hi @birdayz -- thanks for your reply.
1. I agree, but I was unable to find a way to block until the connection was fully established.  If you have any suggestions, I'm happy to discuss them.  Again, my initial thought is that the `ConsumerGroup` interface seems to be easier to use for this purpose as there are interface methods that are called throughout the setup lifecycle.  Checkout the example implementation [here](https://github.com/Shopify/sarama/blob/master/examples/consumergroup/main.go).
2. For interleaved output, the lock was only held at the end when writing out the payload, but not when writing the headers or errors -- for instance in [this block](https://github.com/infinimesh/kaf/blob/master/cmd/kaf/consume.go#L126-L145).  So the headers and errors would be written without having acquired the mutex and thus be interleaved.  To fix this with mutexes, you'd essentially have to hold the lock for the duration of the loop iteration within the goroutine meaning that only one of the inner loops would be running at a time, negating the parallelism of the multiple goroutines consuming partitions.  Channels seem like a cleaner solution to me (with the exception of @mafredri's comment above that should be addressed).

I'll continue to look into the `ConsumerGroup` interface as that might clean things up significantly.

**Comment by mafredri** [comment]
> 2. For interleaved output, the lock was only held at the end when writing out the payload, but not when writing the headers or errors -- for instance in this block. So the headers and errors would be written without having acquired the mutex and thus be interleaved. [...]

That should not be the case, in the section you linked to, the [tabwriter](https://github.com/infinimesh/kaf/blob/50dd2e49233dab61836bf562ea4bbfb5a7347a5c/cmd/kaf/consume.go#L124) is writing to `stderr` (a `bytes.Buffer`) and as such any output should be delayed until we acquire the mutex lock.

**Comment by sergiosalvatore** [comment]
@mafredri -- I think you're right -- I was just misreading the output.  :-(

**Comment by birdayz** [comment]
1) i'll try to look into it the evening, maybe there's a way to block so long. Regarding the consumer groups: i'd rather not use it, otherwise we have to create new consumer groups every time the application is started. some CLIs do this (actually just one shitty GUI tool i saw, i dont remember the name), but this pollutes the broker with lots of useless consumer groups. consumer groups are the tool of choice if you want to (persistently) let the broker control partition assignment & offsets..which is not the case here.

**Comment by sergiosalvatore** [comment]
Closing this for now...

**Pull request closed without merging**